javalib/src/main/scala/java/lang/Long.scala

```diff
@@ -36,81 +47,345 @@ object Long {
+  }
 
-  def parseUnsignedLong(s: String, radix: Int): scala.Long = ???
+  @inline def compare(x: scala.Long, y: scala.Long): scala.Int =
+    if (x == y) 0
+    else if (x < y) -1
+    else 1
 
-  @inline def parseUnsignedLong(s: String): scala.Long =
-    parseUnsignedLong(s, 10)
+//  @inline def compareUnsigned(x: scala.Long, y: scala.Long): scala.Int = ???
+
+  def decode(nm: _String): Long = {
+    val length = nm.length()
+    if (length == 0) {
+      throw new NumberFormatException()
+    } else {
+      var i          = 0
+      var firstDigit = nm.charAt(i)
+      val negative = firstDigit == '-'
+
+      if (negative) {
+        if (length == 1) {
+          throw new NumberFormatException(nm)
+        }
+
+        i += 1
+        firstDigit = nm.charAt(i)
+      }
+
+      var base = 10
+      if (firstDigit == '0') {
+        i += 1
+        if (i == length) {
+          return valueOf(0L)
+        }
+
+        firstDigit = nm.charAt(i)
+        if (firstDigit == 'x' || firstDigit == 'X') {
+          if (i == length) {
+            throw new NumberFormatException(nm)
+          }
+          i += 1
+          base = 16
+        } else {
+          base = 8
+        }
+      } else if (firstDigit == '#') {
+        if (i == length) {
+          throw new NumberFormatException(nm)
+        }
+        i += 1
+        base = 16
+      }
+
+      valueOf(parse(nm, i, base, negative))
+    }
+  }
 
-  def parseUnsignedLongInternal(
-      s: String, radix: Int, start: Int): scala.Long = ???
+  @inline
+  def divideUnsigned(dividend: scala.Long, divisor: scala.Long): scala.Long =
+    divULong(dividend, divisor)
 
-  @inline def valueOf(longValue: scala.Long): Long = new Long(longValue)
-  @inline def valueOf(s: String): Long             = valueOf(parseLong(s))
+  def getLong(nm: _String): Long = {
+    if (nm == null || nm.length() == 0) {
+      return null
+    }
+
+    val prop = System.getProperty(nm)
+    if (prop == null) {
+      return null
+    }
+
+    try {
+      decode(prop)
+    } catch {
+      case e: NumberFormatException =>
+        null
+    }
+  }
 
-  @inline def valueOf(s: String, radix: Int): Long =
-    valueOf(parseLong(s, radix))
+  def getLong(nm: _String, v: scala.Long): Long = {
+    if (nm == null || nm.length() == 0) {
+      return valueOf(v)
+    }
+
+    val prop = System.getProperty(nm)
+    if (prop == null) {
+      return valueOf(v)
+    }
+
+    try {
+      decode(prop)
+    } catch {
+      case e: NumberFormatException =>
+        valueOf(v);
+    }
+  }
 
   @inline def hashCode(value: scala.Long): Int =
     value.toInt ^ (value >>> 32).toInt
 
-  @inline def compare(x: scala.Long, y: scala.Long): scala.Int = {
-    if (x == y) 0
-    else if (x < y) -1
-    else 1
+  def highestOneBit(i: scala.Long): scala.Long = {
+    var lng = i
+    lng |= (lng >> 1)
+    lng |= (lng >> 2)
+    lng |= (lng >> 4)
+    lng |= (lng >> 8)
+    lng |= (lng >> 16)
+    lng |= (lng >> 32)
+    lng & ~(lng >>> 1)
```

Comment:
This one is better off being a sequence of vals, that incrementally compute the result.